### ISSUE TITLE
TypeScript compiler gives incorrect error for large extension chain

### ISSUE THREAD

**Issue by Jameskmonger** [microsoft/TypeScript#12100]
**TypeScript Version:**  1.8.10

**Code**

```ts
export class BaseModel
{
    public someBaseModeloMethod(): void { }
}

export class BaseView
{
    public someBaseViewMethod(): void { }
}

export class CollectionView<TChildView extends BaseView, TModel extends BaseModel>
    extends BaseView
{
    public someCollectionViewMethod(): void { }
}

export class ItemView<TModel extends BaseModel>
    extends BaseView
{
    public setItem(model: TModel): void { }
}

export interface IGridView<TChildView extends ItemView<TModel>, TModel extends BaseModel>
    extends CollectionView<TChildView, TModel>
{
    someGridViewMethod(): void;
}

export class GridView<TChildView extends ItemView<TModel>, TModel extends BaseModel>
    extends CollectionView<TChildView, TModel>
    implements IGridView<TChildView, TModel>
{
    public someGridViewMethod(): void { }
}

function bind<TInterface>(key: string, implementation: new (...args: any[]) => TInterface) {
    // do some bindings here
}

bind<IGridView<ItemView<BaseModel>, BaseModel>>("IGridView", GridView);
```

**Expected behavior:**

The code should compile correctly, because `GridView` is a constructor for an `IGridView<ItemView<BaseModel>, BaseModel>`

**Actual behavior:** 

The `GridView` part of the final line (`bind` statement) gives me the following error:

    TS2345 Argument of type 'typeof GridView' is not assignable to parameter of type 'new (...args: any[]) => IGridView<ItemView<BaseModel>, BaseModel>'.
    	Type 'CollectionView<any, any>' is not assignable to type 'IGridView<ItemView<BaseModel>, BaseModel>'.
    		Property 'someGridViewMethod' is missing in type 'CollectionView<any, any>'.

As you can see, it is expecting `someGridViewMethod` to be on `CollectionView`. However, `GridView` **extends** `CollectionView`, so why is it expecting to find `GridView` methods on the `CollectionView` class?

**Comment by Jameskmonger** [comment]
I've updated my question - this works fine in `2.0.3` but not in `1.8.10`.

**Comment by RyanCavanaugh** [comment]
Seems to have been fixed
